panel/chat/step.py

```diff
@@ -0,0 +1,239 @@
+
+    success_title = param.String(default=None, doc="""
+        Title to display when status is success.""")
+
+    default_avatars = param.Dict(
+        default=DEFAULT_STATUS_AVATARS,
+        doc="Mapping from status to default status avatar.")
+
+    default_title = param.String(
+        default="",
+        doc="The default title to display if the other title params are unset.")
+
+    failed_title = param.String(
+        default=None,
+        doc="Title to display when status is failed.")
+
+    margin = param.Parameter(
+        default=(5, 5, 5, 10), doc="""
+        Allows to create additional space around the component. May
+        be specified as a two-tuple of the form (vertical, horizontal)
+        or a four-tuple (top, right, bottom, left).""")
+
+    pending_title = param.String(
+        default=None,
+        doc="Title to display when status is pending."
+    )
+
+    running_title = param.String(
+        default=None,
+        doc="Title to display when status is running."
+    )
+
+    status = param.Selector(
+        default="pending", objects=["pending", "running", "success", "failed"])
+
+    title = param.String(default="", constant=True, doc="""
+        The title of the chat step. Will redirect to default_title on init.
+        After, it cannot be set directly; instead use the *_title params.""")
+
+    _rename: ClassVar[Mapping[str, str | None]] = {
+        "collapsed_on_success": None,
+        "default_avatars": None,
+        "default_title": None,
+        "pending_title": None,
+        "running_title": None,
+        "success_title": None,
+        "failed_title": None,
+        "status": None,
+        **Card._rename,
+    }
+
+    _stylesheets: ClassVar[list[str]] = [f"{CDN_DIST}css/chat_step.css"]
+
+    def __init__(self, *objects, **params):
+        self._instance = None
+        self._avatar_placeholder = Placeholder(css_classes=["step-avatar-container"])
+
+        if params.get("title"):
+            if params.get("default_title"):
+                raise ValueError("Cannot set both title and default_title.")
+            params["default_title"] = params["title"]
+
+        super().__init__(*objects, **params)
+        self._render_avatar()
+        self._title_pane = HTML(
+            self.param.title,
+            margin=0,
+            css_classes=["step-title"],
+        )
+        self.header = Row(
+            self._avatar_placeholder,
+            self._title_pane,
+            stylesheets=self._stylesheets + self.param.stylesheets.rx(),
+            css_classes=["step-header"],
+        )
+    def __enter__(self):
+        self.status = "running"
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is not None:
+            self.status = "failed"
+            if self.failed_title is None:
+                # convert to str to wrap repr in apostrophes
+                self.failed_title = f"Failed due to: {str(exc_value)!r}"
```

Comment:
After the first time the failed_title will be set and therefore not be updated if the context manager is invoked a second time.